### PR TITLE
NPC navigation

### PR DESCRIPTION
**Pull request by jfcarter**
This app allows you to have any model go between locations in your domain while cycling through animations.
NOTE: Models and animations must have matching skeletons (i.e. An animation meant for an octopus will not work for a human model.)
Click Edit -> Open and run script from URL -> then enter this link to add the app to your tablet
https://hifi-content.s3.amazonaws.com/jedon/Game_Creater_Toolkit/AI_Nav/AI_Navigation.js
1.	Download NPC_Nav app through file AI_Navigation.js (app named "NPC NAV" with a GPS icon)
2.	Open the app.
3.	Place at least two beacons (max 8) with the "PLACE BEACON" button in order to establish a path for your model.
4.	Enter your own model with associated Idle and Moving animations. Also input the speed and idle time for your model when going from beacon to beacon. If you dont have any models with animations then press "SAMPLE ANIMATIONS" and you will be given a default power ranger model with animations.
5.	When previous are completed press "PREVIEW" to see how it will look.
6.	Adjust the position of your beacons if you arent happy with your preview models route and the model will automatically adjust to the new beacon position.
7.	Press "DONE" and the model will return to the first beacon you placed. Once there it will begin its route.
8. Test error cases, error messages shown in the tablet app. If your beacon number is less than 2 you should see "WARNING: MUST HAVE AT LEAST TWO BEACONS PLACED" and if beacon number is greater than 8 you should see "WARNING: CAN NOT PLACE MORE THAN 8 BEACONS". If you haven't entered any model or animation info then you should see "WARNING: NOT A VALID PATH"

### CODE REVIEW

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    //holds the information of the event recieved
+    var info;
+    //distance between beacon and model
+    var dist;
+    //dimensions of the model
+    var modelDim;
+
+    // "Install" your cool new app to the tablet
+    // The following lines create a button on the tablet's menu screen
+    var button = tablet.addButton({
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: Indentation

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    icon: APP_ICON,
+    text: APP_NAME
+    });
+
+    // When user click the app button, we'll display our app on the tablet screen
+    function onClicked() {
+    tablet.gotoWebScreen(APP_URL);
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: Indentation

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    // Every great app starts with a great name (keep it short so that it can fit in the tablet button)
+    var APP_NAME = "NPC NAV";
+    // Link to your app's HTML file
+    var APP_URL = Script.resolvePath("./Tablet_Navigation_App.html?" + Date.now());
+    //Link to Navigation ICON
+    var APP_ICON = Script.resolvePath("./npc-nav-i.svg");
+    // Get a reference to the tablet
+    var tablet = Tablet.getTablet("com.highfidelity.interface.tablet.system");
+    //holds # of beacons
+    var beaconNum = 0;
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: Per style guide, this should be named numBeacons

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    //holds id of the model
+    var model;
+    //decides whether the object can move or not
+    var move = false;
+    //holds location of next beacon for movement
+    var next = 0;
+    //holds the entities speed
+    var speed;
+    //holds the time the idle animation plays between beacons
+    var Idle_time;
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: These variables should follow the style guidelines

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    var Idle_Anim;
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: Don't abbreviate in variable names

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+            move = true;
+            var modelProperties = {
+                animation: {
+                    url: info.Moving_Animation_URL
+                },
+                modelURL: info.Model_URL
+            };
+            Entities.editEntity(model, modelProperties);
+        } else if ((info.type =="Preview") && ((beaconNum > 8) || (beaconNum < 2))) {
+            Window.alert("Must be 2-8 beacons");
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: This shouldn't be Window.alert - this should appear in the tablet directly, not as part of a separate UI window that uses the desktop UI

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        } else if((info == "DONE") && model == undefined) {
+            print("didnt meet criteria, please fill in all spaces provided.");
+        }
+    }
+    tablet.webEventReceived.connect(onWebEventReceived);
+
+    function MakeMove() {
+        //makes the model move from beacon to beacon
+        if ((next != beaconNum) && (beaconNum >= 2) && (model != undefined) && (move)) {
+            try {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: This whole block needs to be indented

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+                    Entities.deleteEntity(beacons[0]);
+                    beacons = [];
+                    Entities.deleteEntity(model);
+                    model = undefined;
+                }
+            }
+        }
+
+        //makes the model move from beacon to beacon
+        if(move) {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: space after if

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+            //make object go to first beacon placed if its currently at the last beacon
+            if((next) == beaconNum)
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: space after if

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+            {
+                next = 0;
+            //will be else if distance is less than some number I will change target to next beacon
+            } else if(dist < .25) {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: space after if

File: Marketplace/AI_Nav/ConstantSpeed.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+	        	x: 0,
+	        	y: 0,
+	        	z: 0
+	        }
+	    };
+    	Entities.editEntity(_this.entityID, setSpeed);
+    }, 100);
+
+    Entities.deletingEntity.connect(function(entityID){
+	    try{
```
Comment (misslivirose) [inline Marketplace/AI_Nav/ConstantSpeed.js]: space after try

File: Marketplace/AI_Nav/Movement.js [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+//  Copyright 2017 High Fidelity, Inc.
+//
+//  Server script that makes my entity move from position to position based off the beacon locations created with AI_Navigation.js
+//
+//  Distributed under the Apache License, Version 7.1.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+(function() {
+    //holds # of beacons
+    var beaconNum = 0;
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Movement.js]: numBeacons and other variable naming

File: Marketplace/AI_Nav/Movement.js [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    function getDistance(dx, dy, dz) {
+        //get distance between model and beacon
+        var dist = Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2) + Math.pow(dz, 2));
+        return dist;
+    }
+
+    Entities.deletingEntity.connect(function(entityID){
+        try{
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Movement.js]: space after try

File: Marketplace/AI_Nav/Tablet_Navigation_App.html [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                var id = form.Anim_Idle.value;
+                var check2 = id.substr(id.length - 3);
+                var id = form.Idle_Time.value;
+                var check3 = id;
+                var id = form.Speed.value;
+                var check4 = id;
+                var id = form.Model.value;
+                var check5 = id.substr(id.length - 3);;
+
+                if (((check1 == ("fbx")) || (check1 == ("obj")) || (check1 == ("FBX"))) && ((check2 == ("fbx")) || (check2 == ("obj")) || (check2 == ("FBX"))) && (!(isNaN(check3))) && (!(isNaN(check4))) && ((check5 == ("fbx")) || (check5 == ("obj")) || (check5 == ("FBX")))) {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Tablet_Navigation_App.html]: can this be split into separate lines?

File: Marketplace/AI_Nav/Tablet_Navigation_App.html [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+                <input name= "Speed" type="text" placeholder="Speed" class="textboxclass" id="Speed">
+                <p></p>
+            <input type="button" id="Submit" value="Preview" onclick="results(this.form)">
+            <input type="button" id="Sample" value="Sample Animations" onclick="fillValuesInTextBoxes()">
+            </form>
+            <p><input type="button" class="done-button red" value="DONE"></p>
+        </div>
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
+        <script>
+            var beaconNum = 0;
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Tablet_Navigation_App.html]: numBeacons

File: Marketplace/AI_Nav/Tablet_Navigation_App.html [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+            $(".done-button").click(function() {
+                beaconNum = 0;
+                var end = "DONE";
+                EventBridge.emitWebEvent(JSON.stringify(end));
+            });
+
+            $(document).ready(main);
+        </script>
+    </body>
+</html>
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Tablet_Navigation_App.html]: newline EOF

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+    }
+
+    function getDistance(dx, dy, dz) {
+        //get distance between model and beacon
+        var dist = Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2) + Math.pow(dz, 2));
+        return dist;
+    }
+
+    // Helper function that gives us a position right in front of the user 
+    function getPositionToCreateEntity() {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: Indentation should be 4 spaces, not 2

File: Marketplace/AI_Nav/Movement.js [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+        }
+    }, 50);
+
+    function Wait() {
+        //Make model move again
+        move = true;
+        //Use timeout to make lag less noticable
+        Script.setTimeout( function() {
+        //change to its moving animation. 
+        var modelProperties = {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Movement.js]: Indentation

File: Marketplace/AI_Nav/Tablet_Navigation_App.html [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+                var id = form.animIdle.value;
+                var check2 = id.substr(id.length - 3);
+                var id = form.idleTime.value;
+                var check3 = id;
+                var id = form.speed.value;
+                var check4 = id;
+                var id = form.model.value;
+                var check5 = id.substr(id.length - 3);;
+
+                if (((check1 == ("fbx")) || (check1 == ("obj")) || (check1 == ("FBX"))) && ((check2 == ("fbx")) || (check2 == ("obj")) || (check2 == ("FBX"))) && (!(isNaN(check3))) && (!(isNaN(check4))) && ((check5 == ("fbx")) || (check5 == ("obj")) || (check5 == ("FBX")))) {
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Tablet_Navigation_App.html]: Can this line be broken up?

File: Marketplace/AI_Nav/Tablet_Navigation_App.html [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+                        "idleAnimationURL": form.animIdle.value,
+                        "movingAnimationURL": form.animMov.value,
+                        "idleTime": form.idleTime.value,
+                        "speed": form.speed.value
+                    };
+                    // The event bridge handles event represented as a string the best. So here we first create a Javascript object, then convert to stirng
+                    EventBridge.emitWebEvent(JSON.stringify(readyEvent));
+                } else {
+                    if (numBeacons < 2) {
+                        window.alert("must have at least two beacons placed");
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Tablet_Navigation_App.html]: Don't use Window.alert here

File: Marketplace/AI_Nav/Tablet_Navigation_App.html [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+                        window.alert("not a valid path");
+                    }
+                }
+            }
+
+            // Send an event when user click on each of the gemstone buttons
+            $(".beacon-button").click(function() {
+                numBeacons++;
+                if (numBeacons == 9) {
+                    window.alert("Too many beacons");
```
Comment (misslivirose) [inline Marketplace/AI_Nav/Tablet_Navigation_App.html]: Don't use window.alert

File: Marketplace/AI_Nav/AI_Navigation.js [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+            move = true;
+            var modelProperties = {
+                animation: {
+                    url: info.movingAnimationURL
+                },
+                modelURL: info.modelURL
+            };
+            Entities.editEntity(model, modelProperties);
+        } else if ((info.type =="Preview") && ((numBeacons > 8) || (numBeacons < 2))) {
+            Window.alert("Must be 2-8 beacons");
```
Comment (misslivirose) [inline Marketplace/AI_Nav/AI_Navigation.js]: Don't use Window.alert  to display information outside of the main UI

File: Marketplace/AI_Nav/AI_Navigation.js (hunk 17 above)
Comment (jfcarter) [reply Marketplace/AI_Nav/AI_Navigation.js]: Little confused on this one. Where is it two spaces?

Comment (jfcarter) [reply Marketplace/AI_Nav/AI_Navigation.js]: Nevermind I see what you mean now

**Comment by ShoaibSmith** [comment]
Is there a build for this?

**Comment by skytoad** [comment]
Testing

**Comment by skytoad** [comment]
Overall pass, some issues. 

1. Download NPC_Nav app through file AI_Navigation.js (app named "NPC NAV" with a GPS icon) **Pass.**

2. Open the app. **Pass**

3. Place at least two beacons (max 8) with the "PLACE BEACON" button in order to establish a path for your model. **Pass**

4. Enter your own model with associated Idle and Moving animations. Also input the speed and idle time for your model when going from beacon to beacon. If you dont have any models with animations then press "SAMPLE ANIMATIONS" and you will be given a default power ranger model with animations. **Pass**

5. When previous are completed press "PREVIEW" to see how it will look. **Pass**

6. Adjust the position of your beacons if you arent happy with your preview models route and the model will automatically adjust to the new beacon position. **Pass**

7. Press "DONE" and the model will return to the first beacon you placed. Once there it will begin its route. **Pass**

8. Test error cases, error messages shown in the tablet app. If your beacon number is less than 2 you should see "WARNING: MUST HAVE AT LEAST TWO BEACONS PLACED" and if beacon number is greater than 8 you should see "WARNING: TOO MANY BEACONS". If you haven't entered any model or animation info then you should see "WARNING: NOT A VALID PATH." **If you click preview without inputting avatar information, the error message will show "must have at least two beacons placed" even if you do in fact have two beacons placed. Inputting avatar data and then placing another beacon resolves this issue,**

**Comment by skytoad** [comment]
Pass without issues.

**Pull request merged**